### ISSUE TITLE
Inconsistent behavior in python-mypy (invalid errors shown)

### ISSUE THREAD

**Issue by vikigenius** [flycheck/flycheck#1711]
## Checklist
-  I have checked existing issues for potential duplicates before creating this one.
-  I have read the [Troubleshooting guide][].

## Bug description
flycheck python-mypy produces errors that are not valid/not observed when I manually run it from the command line

## Steps to reproduce
Here is a sample code snippet that is pretty common in sqlalchemy

```
from sqlalchemy import Column, Integer, String
from sqlalchemy.ext.declarative import declarative_base

Base = declarative_base()


class User(Base):
    """The user model."""

    __tablename__ = 'user'
    uid = Column(Integer, primary_key=True)
    name = Column(String)

```

1. Errors are shown for the line `User(Base):`
2. run flycheck-compile python-mypy on the buffer
```
mypy --show-column-numbers /home/void/Projects/redorg/redorg/user.py
user.py:9:12: error: Variable "redorg.user.Base" is not valid as a type
user.py:9:12: error: Invalid base class "Base"
Found 2 errors in 1 file (checked 1 source file)
```
2. Run the same command in the command line, and it doesn't produce any errors.

## Expected behavior
flycheck should not show any errors

## System configuration
```
Syntax checkers for buffer user.py in python-mode:

First checker to run:

  python-flake8
    - may enable:         yes
    - executable:         Found at /home/void/.miniconda3/envs/redorg/bin/flake8
    - configuration file: Not found
    - next checkers:      python-pylint, python-mypy

Checkers that may run as part of the first checker's chain:

  python-mypy
    - may enable:         yes
    - predicate:          t
    - executable:         Found at /home/void/.miniconda3/envs/redorg/bin/mypy
    - configuration file: Not found

Checkers that could run if selected:

  python-pycompile  select
    - may enable:    yes
    - executable:    Found at /home/void/.miniconda3/envs/redorg/bin/python3
    - next checkers: python-mypy

Checkers that are compatible with this mode, but will not run until properly configured:

  python-pylint (disabled)
    - may enable:         Automatically disabled!
    - executable:         Not found
    - configuration file: Not found
    - next checkers:      python-mypy

Flycheck Mode is enabled. Use SPC u C-c ! x to enable disabled checkers.

--------------------

Flycheck version: 32snapshot
Emacs version:    26.3
System:           x86_64-unknown-linux-gnu
Window system:    x
```

Emacs configuration:
-  Doom Emacs

## Additional notes

[Troubleshooting guide]:  https://www.flycheck.org/en/latest/user/troubleshooting.html

**Comment by vikigenius** [comment]
After further debugging, I have managed to isolate the issue better. I use a mypy plugin: sqlalchemy-stubs: https://github.com/dropbox/sqlalchemy-stubs/

My setup.cfg has a line in the `[mypy]` section: `plugins = sqlmypy`

If i remove this line, the command line execution produces the same errors. This means that, flycheck when running the mypy checker somehow ignores this plugin specification.

**Comment by cpitclaudel** [comment]
Could it be a working directory issue ? When you say `Run the same command in the command line, and it doesn't produce any errors.`, which directory are you running this from?

**Comment by vikigenius** [comment]
@cpitclaudel Yes I run it from my project root where the setup.cfg is located.

I believe mypy works only when run from the project root. https://github.com/python/mypy/issues/3377

**Comment by vikigenius** [comment]
lsp people have seemingly run into the same problem, doesn't appear as though they have fixed it. https://github.com/tomv564/pyls-mypy/issues/9

I don't know if it is possible to always run the checkers from the project root?

**Comment by vikigenius** [comment]
Following this comment https://github.com/flycheck/flycheck/issues/721#issuecomment-129556743 I created a dir-locals in my project root letting flycheck know where my setup.cfg is located: `((python-mode . ((flycheck-python-mypy-ini . "setup.cfg"))))` . The issue is resolved. I don't know if the behavior currently is considered a bug or not.

**Comment by cpitclaudel** [comment]
> I don't know if the behavior currently is considered a bug or not.

Well, it's a problem, at least.

> I don't know if it is possible to always run the checkers from the project root?

The difficulty is identifying the project root.  How are we supposed to tell?  Currently we look for mypy.ini only.  We could look for other names too.

**Comment by vikigenius** [comment]
@cpitclaudel From the docs, just mypy.ini and setup.cfg should do: https://mypy.readthedocs.io/en/latest/config_file.html

**Comment by cpitclaudel** [comment]
I looked into this.  It's a bit trickier than it might seem.

The reason why setting `flycheck-mypy-python-ini` works in your case is that mypy seems to pick up the hint about the project root when passed the path to a config file (even without being run from the right project directory).  Can you check how mypy is called with `C-c ! C-c` when you specify the ini file as `setup.cfg`?

So we have basically two options:

- Generalize the config file setting it to allow multiple files (so we would change the default to `mypy.ini` and `setup.cfg`), and passing whatever we find to mypy.

- Search for config files, use that as the root folder for mypy and flake8.  We need to decide how we search for these config files though. Do we do the full flycheck-locate-config-file logic, or something simpler? And how does that interact with the config-file setting?

I'd love to hear opinions about that.

**Comment by vikigenius** [comment]
@cpitclaudel I just checked

```
-*- mode: compilation; default-directory: "~/Projects/redorg/redorg/" -*-
Compilation started at Thu Mar 26 01:43:41

mypy --show-column-numbers --config-file /home/void/Projects/redorg/setup.cfg /home/void/Projects/redorg/redorg/models.py
Success: no issues found in 1 source file

Compilation finished at Thu Mar 26 01:43:41
```
This doesn't make it very clear, it seems to still be executing from the wrong directory but mypy maybe automagically figures out the right directory from the config option ?

**Comment by cpitclaudel** [comment]
Thanks.  So maybe what we need is to make the code that locates the config file smarter, rather than setting a specific working directory.

**Comment by cpitclaudel** [comment]
Can you try the branch at https://github.com/flycheck/flycheck/tree/cpitclaudel_fix-1711-mypy-multiple-config-files ? I think it should work regardless of what you set the flycheck-python-mypy-ini variable to now.

**Comment by vikigenius** [comment]
@cpitclaudel Yes, it works without the dir-locals.el now

**Comment by cpitclaudel** [comment]
Thanks!

**Comment by lbolla** [comment]
See this commit https://github.com/lbolla/emacs-flycheck-mypy/commit/d5c791f where setting `working-dir` was added to the now deprecated `emacs-flycheck-mypy` library (merged to Flycheck some time ago).

Maybe we can port that patch to the Mypy check in Flycheck? Do you want me to issue a PR?

**Comment by cpitclaudel** [comment]
> Maybe we can port that patch to the Mypy check in Flycheck? Do you want me to issue a PR?

I think it's unnecessary: the approach we took in #1714 makes it unnecessary to change the working directory.
